python/sglang/lang/interpreter.py

```diff
@@ -109,63 +110,63 @@ def run_program_batch(
 
+    rets = []
     if num_threads == 1:
-        rets = []
-        if progress_bar:
-            for arguments in tqdm.tqdm(batch_arguments):
-                rets.append(
-                    run_program(
-                        program,
-                        backend,
-                        (),
-                        arguments,
-                        default_sampling_para,
-                        False,
-                        True,
-                    )
-                )
+        if progress_bar and not generator_style:
+            iterator = tqdm.tqdm(batch_arguments)
         else:
-            for arguments in batch_arguments:
-                rets.append(
-                    run_program(
-                        program,
-                        backend,
-                        (),
-                        arguments,
-                        default_sampling_para,
-                        False,
-                        True,
-                    )
-                )
+            iterator = batch_arguments
+
+        for arguments in iterator:
+            result = run_program(
+                program,
+                backend,
+                (),
+                arguments,
+                default_sampling_para,
+                False,
+                True,
+            )
+            if generator_style:
+                yield result
+            else:
+                rets.append(result)
+
     else:
-        if progress_bar:
+        if progress_bar and not generator_style:
             pbar = tqdm.tqdm(total=len(batch_arguments))
 
         with ThreadPoolExecutor(num_threads) as executor:
             futures = []
+            future_to_arguments = {}
             for arguments in batch_arguments:
-                futures.append(
-                    executor.submit(
-                        run_program,
-                        program,
-                        backend,
-                        (),
-                        arguments,
-                        default_sampling_para,
-                        False,
-                        True,
-                    )
+                future = executor.submit(
+                    run_program,
+                    program,
+                    backend,
+                    (),
+                    arguments,
+                    default_sampling_para,
+                    False,
+                    True,
                 )
-                if progress_bar:
-                    futures[-1].add_done_callback(lambda _: pbar.update())
-
-            rets = [f.result() for f in futures]
-        rets[-1].sync()
+                futures.append(future)
+                future_to_arguments[future] = arguments
+                if progress_bar and not generator_style:
+                    future.add_done_callback(lambda _: pbar.update())
+
+            if generator_style:
+                for future in concurrent.futures.as_completed(futures):
```

Comment:
This changes the order of returned results. This might not be expected. I think we should at least explicitly say this or use two flags to control this behavior. What do you think?